### ISSUE TITLE
Wavelets applying only partially

### ISSUE THREAD

**Issue by luisfl** [Beep6581/RawTherapee#3793]
After I processed the attached ORF file and saved it as TIFF,
the image in RT window looks more aggressively sharpened than the one saved (see crops attached taken as screenshots).
I then realized that the wavelet local contrast was applied only 'partially' in the saved version: if I disable completely the wavelet tab, the two images are identical.

Files attached in [report.zip](https://github.com/Beep6581/RawTherapee/files/887301/report.zip): ORF, crops and pp3 file.

RT Version: rawtherapee-5.0.r1-7.1.x86_64
OS: Fedora 23 64bits fully updated

**Comment by TooWaBoo** [comment]
Could not reproduce on Windows.
How did you save the image? Fast Export, Queue or Save as?

```
Version: 5.0-r1-gtk3-335-gcd93dab9
Branch: dev
Commit: cd93dab9
Commit date: 2017-03-30
Compiler: gcc 6.3.0
Processor: undefined
System: Windows
Bit depth: 64 bits
Gtkmm: V3.22.0
Build type: release
```

**Comment by luisfl** [comment]
Saved as "Save as". But I realized something wrong was going on right when I saw the image in GIMP sent by the external editor button, before saving (I never save directly from RT anyway). So neither is working properly.

**Comment by TooWaBoo** [comment]
Bug confirmed.
I didn't see it the first time. So I overdo Wavelet -> Contrast. (see screenshot). The Contrast setttings are ignored, not only in Tiff also in JPG.

![unbenannt](https://cloud.githubusercontent.com/assets/13800920/24580711/f2c94790-170c-11e7-9ce2-151eb13b5ef1.png)

**Comment by luisfl** [comment]
Actually, it has nothing to do with the Saving process, since the problem is present even with the "edit in external editor" button.

**Comment by TooWaBoo** [comment]
"edit in external editor" saves the image in a temp folder and then opens the editor. Looks like "Contrast" is missing in the save processing pipeline.

**Comment by luisfl** [comment]
I see.
I am also trying the unstable branch

rawtherapee-unstable-5.0.r1~git.20170401.335.cd93dab9-1.2.x86_64

It has the very same bug, but also with Shadow-Highlights and Tone mapping tools.
Don't know if the stable version also omits these tools.

L.

**Comment by heckflosse** [comment]
@luisfl For tonemapping please take a look [here](http://rawpedia.rawtherapee.com/Tone_Mapping#Getting_the_preview_to_match_the_saved_image)

**Comment by luisfl** [comment]
The stable branch seems to process Tonemapping and Shadow-Highlights just fine, as opposed to the dev version. I'll stick with the stable branch then.

**Comment by heckflosse** [comment]
Can you please post more details about differences between dev and stable concerning Tonemapping and Shadow-Highlights?

Edit: I ask because latest changes to Tonemapping are made about half a year ago at least, which does not fit to differences between current dev and stable (though you didn't mention what the 'stable' version you used is)

**Comment by luisfl** [comment]
The dev version I tried, rawtherapee-unstable-5.0.r1~git.20170401.335.cd93dab9-1.2.x86_64, has the same problem that rawtherapee-5.0.r1-7.1.x86_64 has, but with Tonemapping and Shadow-Highlights instead of wavelets tab: these two tools are not applied before saving the file, so the saved version looks quite different than the one showed on RT window. It is not a problem on the algorithms or whatever, it's just that they are not applied.

**Comment by heckflosse** [comment]
@luisfl I just tried with latest dev (cd93dab9). Both tools are applied before saving the file (in preview) here.

**Comment by heckflosse** [comment]
@luisfl Anyway. If there is an Issue with this tools (tonemapping or Shadows/Highlights) we (you) should open a new Issue instead of highjacking your own Wavelet issue ;-)

**Comment by luisfl** [comment]
It looks to me like the very same issue, but with different tools. I don't think it is necessary to open a ticket for each tool.

**Comment by heckflosse** [comment]
Ok, here's a screenshot:
Left is neutral profile, right is neutral + Shadows/Highlights. Do you still claim that Shadows/Highlights is not applied in preview (before saving the file)?
![shhl](https://cloud.githubusercontent.com/assets/13733487/24582853/0e714d22-1739-11e7-8a98-5ba4e6cd3c64.png)

**Comment by sguyader** [comment]
Into, I think that what @luisfl is trying to say is that the effect is actually shown in the RT preview, but not in the saved image.

**Comment by luisfl** [comment]
Correct. This is exactly the same problem as with wavelets for the stable release.

**Comment by heckflosse** [comment]
@sguyader Ok, I will check that. But at least for Shadows/Highlights @luisfl said the opposite...

`these two tools are not applied before saving the file`

**Comment by luisfl** [comment]
No. It's always the same problem: file not saved as the RT preview window. Look (identical PP3, only SH and Tonemapping applied:

![lua_jn-170309-a_32082-stable](https://cloud.githubusercontent.com/assets/15525268/24583013/eeb8f6bc-1713-11e7-8b19-951853d88f23.jpg)
![lua_jn-170309-a_32082-unstable](https://cloud.githubusercontent.com/assets/15525268/24583014/eebb18f2-1713-11e7-817d-98b2c40afbf6.jpg)

Stable is the first, unstable is the second. Both RT windows shows the very same image, the one as the stable version.

Therefore, just to clarify: BOTH stable and unstable versions are not saving the complete workflow, leaving some tools unapplied. Just different tools for each version (wavelets for stable, SH and maybe tonemapping for unstable). Yet, the RT window shows the tools applied in both versions. The problem happens while saving the file.

**Comment by heckflosse** [comment]
@sguyader Sébastien, can you confirm the issue?

**Comment by TooWaBoo** [comment]
@luisfl 
Confirmed.
Here's an extrem example with Shadow / Highlight
![unbenannt](https://cloud.githubusercontent.com/assets/13800920/24583045/b78ed5a0-173e-11e7-9798-0cbab804ed6a.PNG)

**Comment by heckflosse** [comment]
@TooWaBoo Ok, that shows a difference. But does it show that Shadows/Highlights is not applied in preview? And does it show that we should handle it in this issue? I don't think so. That was my point.
I'm absolutely not complaining about this 'wavelet' issue and also not about the 'ShHl' issue, but we should handle each in its own issue to avoid confusion (the confusion which already happened by not doing this, which is a waste of our all time)

**Comment by TooWaBoo** [comment]
@heckflosse 
The preview (right) is OK but the saved image (left) is wrong.
It's the same issue like the Wavelet.

**Comment by heckflosse** [comment]
Please continue discussion on Shadows/Highlights [here](https://github.com/Beep6581/RawTherapee/issues/3796)

**Comment by heckflosse** [comment]
@Desmis Jacques, can you also have a look at the Issue with wavelet preview and output?

Ingo

**Comment by heckflosse** [comment]
@luisfl @TooWaBoo @sguyader I think I found it.
Can you please try with 'Fast export' set to 'Standard' ?

Seems to be a fallout of the changes of @agriggio made to 'fast export'

I apologize for that!

Ingo

**Comment by heckflosse** [comment]
Scratch my last comment. There seems to be something else... still searching...

**Comment by heckflosse** [comment]
I'm bisecting

**Comment by heckflosse** [comment]
I bisected. Not to the final commit but close enough. https://github.com/Beep6581/RawTherapee/commit/209ea3533b3176444848314b1ce468dd30db938a works fine but https://github.com/Beep6581/RawTherapee/commit/518d51e8395948622a1dcaf7c991989a3e8992f4 does not.

@agriggio Please take a look!

**Comment by agriggio** [comment]
i am looking already. wavelets are applied also when saving, but the result
is different from the preview. I am trying to understand why

**Comment by agriggio** [comment]
btw, @luisfl said he saw this already in an official release, correct? that
should exclude the fast export stuff... anyway, working on it

**Comment by heckflosse** [comment]
@agriggio try wit Shadows/Highlights. It's not applied when you save with https://github.com/Beep6581/RawTherapee/commit/518d51e8395948622a1dcaf7c991989a3e8992f4

**Comment by agriggio** [comment]
I just tried with `209ea35`, but I still see the wavelet issue. Here's a screenshot. I used the pp3 of @luisfl but I maxed out the wavelet contrast to make it clearer.
![comparison](https://cloud.githubusercontent.com/assets/11982086/24585363/b1d0a96e-1788-11e7-8c7e-3f49b9f222f4.png)

**NOTE**: I haven't looked at shadows/highlights. I thought this was a separate issue?

**Comment by agriggio** [comment]
I confirm the issue with shadows/highlights. I will look into that then, since it seems related to my stuff

**Comment by Desmis** [comment]
I try with "moon.orf"...and pp3 joined
I don't see difference between preview and Output TIFF (export in Photoshop)... :)

**Comment by agriggio** [comment]
I found the problem with shadows/highlights. I am building `dev` to test it there, and then I'll push

**Comment by agriggio** [comment]
shadows/highlights should work now. But the problem with wavelets seems to be a bit deeper. I checked and in both cases (preview and export) the wavelet parameters are the same. Could it be that the two pipelines are slightly different and we just hit a corner case in which the difference is very visible?

**Comment by heckflosse** [comment]
@agriggio Fix for Shadows/Highlights confirmed. Thank you :+1: 

For wavelets the pipelines are not different but there is a part which evaluates the image which in full processing is the whole image but in preview is only a part of the image.
It's [this part](https://github.com/Beep6581/RawTherapee/blob/dev/rtengine/ipwavelet.cc#L1390).
This can only be solved if we use the values from improccoordinatior.cc ip_wavelet call (where we kind of have the full image (though it's scaled it should reflect the whole image much better than a crop) and then we could pass this values to the ipwavelet_call in dcrop.cc
But that's a bit of work. I can take that part if you like.

**Comment by agriggio** [comment]
@heckflosse I was suspecting something like that, but wasn't sure
since I don't know the wavelets code. if you want to go no problem for me.
alternatively, if you feel like mentoring, I can try to do it with your
advice -- but it might take longer...

**Comment by heckflosse** [comment]
@agriggio The principle should be easy: Pass the evaluated values back from ip_wavelet to improccoordinator.cc and save them in some new instance variables to have them available in dcrop.cc for the ip_wavelet call. Use them there instead of evaluating again.
Do all this only if `params->wavelet.Tilesmethod == "full"`
This 'bug' is so old that it doesn't matter if it takes a bit longer to fix it. Feel free to try it and don't hesitate to ask :+1: 

Edit: I saw there are 2 evaluate2() calls. I guess we have to pass the values of both calls.
Edit2: I guess the same has to be done for `float madL[8][3];`

**Comment by agriggio** [comment]
:+1:, I'll give it a try then!

**Comment by luisfl** [comment]
> @agriggio: i am looking already. wavelets are applied also when saving, but the result is different from the preview. I am trying to understand why

This is correct. I observed that the tool is applied, but only partially, in a "weak" way.

**Comment by heckflosse** [comment]
@agriggio I did something similar for denoise_info some time ago. Search for `denoiseInfoStore`

**Comment by luisfl** [comment]
> @agriggio: btw, @luisfl said he saw this already in an official release, correct? that should exclude the fast export stuff... anyway, working on it

The S/H tool problem was in the stable release. The wavelets, in the unstable. However, I just downloaded the unstable 5.0-r1-gtk3-335-gcd93dab9, and it has problems in the S/H too. So cd93dab9 has problems in both tools: they are applied partially.

Just to be sure, I checked the Tonemapping tool, that I suspected it has problems too, but it works as it should in both the stable and unstable releases.

However, I am a bit worried that two tools have the same problem. I am actually surprised that the problems in S/W-stable and Wavelets-unstable are unrelated. I use very few tools of RT, so it could very well be that other tools have similar problems...

**Comment by heckflosse** [comment]
@luisfl 

> The S/H tool problem is in the stable release. The wavelets, in the unstable.

I expected it to be vice versa...

**Comment by luisfl** [comment]
> @heckflosse: I expected it to be vice versa...

Yes, sorry, you are right. But now the unstable has problems in both tools.

**Comment by heckflosse** [comment]
@luisfl 
@agriggio fixed the shadows/highlights problem in 3a90374d599c56a19155488d0ff312b776f77e7e

**Comment by agriggio** [comment]
@heckflosse I tried to implement your proposed solution, but I'm probably doing something wrong because I see no difference. I'm attaching the current patch (warning: very crude :-)
If you have any feedback, it's more than welcome!
[improccoordinator-save-wavelet-params.patch.txt](https://github.com/Beep6581/RawTherapee/files/888670/improccoordinator-save-wavelet-params.patch.txt)

**Comment by heckflosse** [comment]
@agriggio I will try your patch tomorrow

**Comment by Desmis** [comment]
@agriggio <https://github.com/agriggio>
I just tested your patch :)

As you I see no differences.
Pipelines for RT and wavelets are complex !

Perhaps (I have not verified in detail your patch) others ways :
1) for denoise, we have implanted in dcrop.cc an other complex solution,
but not perfect
2) in this case (the moon and crop), how realy works wavelets : at level 7
it's a square 128 * 128, and background is near to black ?

I w'll look more today
jacques

**Comment by agriggio** [comment]
@Desmis thanks for the feedback! I admit I still don't understand the flow completely, but I'm starting to believe that it will be necessary to work on the whole full-resolution image in order to have an accurate preview for the wavelet tool. Am I wrong?

**Comment by Desmis** [comment]
@agriggio 
Full image ? probably... it's that we done for Retinex !
But is it necessary, and it's complex to do, and w'll slow process. 
jacques

**Comment by agriggio** [comment]
@Desmis yes, I understand your concerns. But I'd say that slow is better than inaccurate, since the whole point of the 1:1 preview is to see accurately what is going on :-) Of course, we should try to minimize the need of this full-processing, and only do it when absolutely necessary. This seems to be the case for the contrast tool in wavelets. Is it true also for other wavelet steps? Also, am I right in assuming this only applies when using the "Full image" method (i.e. no tiling)?

**Comment by Beep6581** [comment]
I use wavelets a lot and have never had an issue with the preview (zoomed to fit) not matching the saved image. I use only the Edge Sharpness, Residual Image and Final Processing sub-tools, so these at least seem to not require full-resolution processing.

**Comment by agriggio** [comment]
I also have never had an issue with wavelet denoising. Sometimes I also use contrast, but never noticed anything strange there as well. That's why I was also suspecting that the picture of @luisfl hits some sort of corner case...

**Comment by heckflosse** [comment]
@agriggio I think I found a bug in your patch.
Shouldn't you skip also the evaluation of madL when evalparams->evaluate == false ?
I mean the loop which starts in line 875 of ipwavelet.cc

Edit: No, you don't have to skip it, but you can skip it.

I think I found another thing we should evaluate only once for the whole image, checking...

**Comment by Desmis** [comment]
Excuse my very bad english, I am an "old french man" :)

I never had problem with that, only in some cases (grey images), the
feeling that there are differences between preview and output . But if I
used "mesure" either with Photoshop (pipette), either with RT on preview
(pipette or Hombre tools), the differences are very tiny or no differences.

In this cas (moon), Is not it just a difference in color management,
increase by black and white  ?

I have very much work on "wavelet for RT" (denoise, wavelet levels, etc.)
and I don't think there is really a big problem, maybe I am wrong, but I
don't think so :)

jacques

**Comment by heckflosse** [comment]
@agriggio line 1710 to 1759 and 1773 to 1775 of ipwavelet.cc should also be evaluated only for whole image. Place a printf("%f\n",ave); in line 1776 and compare the values of 1:1 preview and full processing to see the large differences.

**Comment by heckflosse** [comment]
@Desmis 
Jacques, I also don't think this is a big problem, but if we can solve it we will also save a bit of processing time in preview as a bonus ;-)

**Comment by Desmis** [comment]
Ingo

You are true :)

**Comment by heckflosse** [comment]
@agriggio @Desmis Here's a screen shot (made with current patch) which clearly shows the different output. The area you see is obviously the same in both cases but the detail window output is completely different because it calculates a different average and so on.
![moon00](https://cloud.githubusercontent.com/assets/13733487/24607078/683a57bc-1870-11e7-9688-ac6e7d2d5c9f.png)

To reproduce just use the pp3 from op, place a detail window on the moon and set contrast for level 2 to 350

**Comment by agriggio** [comment]
@heckflosse thanks for the review, I'll revise my patch and upload a new version. Or better, I'll create a branch :-)
@Desmis I haven't seen such big differences in my images, but the screenshot above shows that in extreme cases there can be some significant ones. I don't know what is the exact technical reason for the difference, but the saved moon image seems way less "contrasty" than the preview. As a user, I would find that surprising/annoying...

**Comment by heckflosse** [comment]
@agriggio The screenshot above does not show the saved moon. It shows the 1:1 preview of the raw and the same 1:1 preview area in a detail window. This way you can also test without saving the image ;-)

**Comment by Desmis** [comment]
Ingo

Your screen shot is a good example, very small differences in normal use,
but significantly differences in extreme case :)
Thank you for this highlighting!

jacques

**Comment by luisfl** [comment]
Well... not really for extreme cases only. Of course I didn't have to bump the local contrast so much to realize that the saved version was different from the RT window version. It was just a bit. I was satisfied with the RT version but not with the GIMP version, I wondered why, compared the images, and bingo. Small difference, but enough to not be satisfied.

**Comment by agriggio** [comment]
I just pushed a new branch `wavelet-contrast-accurate-preview` (commit 079c4f2) with an experimental fix for the "bug".
The code now uses a full-res preview in `ImProcCoordinator` if the wavelet contrast tool is enabled. The slowdown is quite visible (but you only pay for it if you use wavelet contrast and ask for a 1:1 preview), but at least what you see should be accurate.
Testing, comments and feedback highly welcome. Note that the code has not been polished at all :-)

**Comment by Desmis** [comment]
@agriggio 

I just tested  this new barnch, the result is quasi perfect :)
There are others features that involve "contrast" in another way : Denoise and refine, Edge sharpness, Final touchup

What to do ?

jacques

**Comment by agriggio** [comment]
@Desmis thanks for testing! in principle we can adapt the same solution to
other wavelet steps if needed. however I would make this an option (under
the performance tab) to let users decide their tradeoff accuracy/cost. for
example for the moon picture already using a 50% zoom produces a close
enough preview for me. what do others think?

**Comment by heckflosse** [comment]
@agriggio Let me try to optimize it. Then we will have a better base for a decision.

**Comment by agriggio** [comment]
@heckflosse sure, go ahead! and thanks :+1:
as usual, my performance metric was the number of lines of the patch ;-)

**Comment by heckflosse** [comment]
;-)

**Comment by heckflosse** [comment]
@agriggio I will try to do that tomorrow but can't promise that

**Comment by agriggio** [comment]
no problem, no hurry for me

**Comment by Desmis** [comment]
@agriggio

I think, since we decided to work in "full image" that we must go to the end of the reasoning.
I think that many users do not use other headings than "contrast", because: 1) the others softwares have only this heading; 2) the documentation - which I have produced - is in French.
But when you master wavelet completely, the rubric that one will probably use least is "contrast".
I would like to put the "full image" processing by default for all the topics (including chroma), and to provide an option to disable
http://rawpedia.rawtherapee.com/Wavelet_levels/fr
This is a very good job

Jacques

**Comment by agriggio** [comment]
@Desmis if you can help me identifying the parameters that affect the other wavelet steps, I can try generalizing the solution used for contrast.
In the meantime, I pushed another changeset to `wavelet-contrast-accurate-preview` that lets the user select the tradeoff accuracy/performance. There's a new option in the performance tab of the preferences dialog, see the screenshot attached.
![wavelet-prefs](https://cloud.githubusercontent.com/assets/11982086/24646327/c8c03146-191b-11e7-8da7-341a5f4c3f11.png)

**Comment by Desmis** [comment]
@agriggio 
These are the same parameters: mean, sigma, etc.
But, with the various possible links: chroma link, denoise and refine link, all modules where decomposition intervenes, are concerned ..
jacques

**Comment by agriggio** [comment]
@Desmis do you have some examples (RAW+pp3) that show a difference in preview vs exported image for the other wavelet steps? That would help me understand what we are looking after... I tried with some extreme settings for edge sharpness but the preview looks accurate to me already (in my branch `wavelet-contrast-accurate-preview`)
![edge-sharpness](https://cloud.githubusercontent.com/assets/11982086/24655717/a326165c-193f-11e7-8cd4-6b93bbe1d7fd.png)

**EDIT**: I forgot to mention, you need to apply this patch to enable the use of the full image also for the other wavelet steps:

```
diff --git a/rtengine/improccoordinator.cc b/rtengine/improccoordinator.cc
--- a/rtengine/improccoordinator.cc
+++ b/rtengine/improccoordinator.cc
@@ -869,7 +869,7 @@
     return (highDetailNeeded
             && options.rtSettings.wavelet_preview_mode != Settings::WAVELET_CROP_100
             && params.wavelet.enabled
-            && params.wavelet.expcontrast
+            //&& params.wavelet.expcontrast
             && params.wavelet.Tilesmethod == "full");
 }
 
```

**Comment by Beep6581** [comment]
Off-topic, but... Could such a solution also be implemented for Tone-mapping? That's one tool which users like to turn up high, and then are disappointed when the saved image looks nothing like the preview.

**Comment by agriggio** [comment]
@Beep6581 I thought that tone mapping was already accurate when the preview is 1:1, am I wrong? Perhaps I never noticed anything strange because although I use tone mapping a lot, I always try to be conservative in my settings... do you have examples of mismatch?

**Comment by Beep6581** [comment]
@agriggio I can confirm that in dev the 100% preview (and 100% detail windows) are accurate, but the preview is not accurate when zoomed-to-fit. It is the zoomed-to-fit preview which is often necessary when using tone mapping.

**Comment by heckflosse** [comment]
@agriggio I found a bug

1) open image and enable wavelets and wavelet contrast => preview is in slow but accurate mode, good!
2) disable wavelet or contrast in wavelet => preview stays in slow but accurate mode.

Another downside of the slow but accurate mode is, that now also per-pixel functions are executed for the whole image which slows them down a lot. Activate the `#define BENCHMARK` in line 43 of improcfun.cc and look at the processing times logged for rgbProc.

Edit: by per-pixel functions I mean functions which do the same calculation for each pixel without influence of the surrounding pixels.

**Comment by agriggio** [comment]
@heckflosse, thanks for testing!

> disable wavelet or contrast in wavelet => preview stays in slow but accurate mode.

what do the console messages say? You should see something like this:

```
SAVING EVALPARAMS!
expecting to use evalparams in load mode
LOADING EVALPARAMS!
expecting to use evalparams in load mode
LOADING EVALPARAMS!
CHANGE FULL RES: 1
SAVING EVALPARAMS!
expecting to use evalparams in load mode
LOADING EVALPARAMS!
expecting to use evalparams in load mode
LOADING EVALPARAMS!
CHANGE FULL RES: 0
```

Where the last line was printed immediately after switching off wavelets... it seems to be working here.

> Another downside of the slow but accurate mode is, that now also per-pixel functions are executed for the whole image which slows them down a lot.

Yes, that's expected. I don't know how to easily work around that... slow really means slow :-)

My current workaround is to use a 50% image instead of a 100% one by setting the new option in "preferences -> performance". The accuracy is still ok for me, and the performance is acceptable (at least on my work laptop -- it will be interesting to see how it goes on my 8 year old thinkpad at home :-)

**Comment by agriggio** [comment]
@Beep6581 

> It is the zoomed-to-fit preview which is often necessary when using tone mapping.

then unfortunately this approach is not applicable in that context I'm afraid. Using the full image when the preview is not 1:1 would slow things down too much IMHO

**Comment by heckflosse** [comment]
@agriggio The last line is indeed `CHANGE FULL RES: 0` but if You change for example exposure after disabling wavelet you will see that it takes longer than in non accurate mode. As I wrote, activate the `#define BENCHMARK` for the timings. You can also see it when panning around because the pan background stays full resolution when the new option in preferences is set to 100%

**Comment by agriggio** [comment]
@heckflosse yes, you are right. Sorry, I didn't check carefully enough :-/ 
Should be fixed now if you update

**Comment by heckflosse** [comment]
@agriggio Fix confirmed :+1:

**Comment by agriggio** [comment]
@Beep6581 for tone-mapping, can the workaround described in rawpedia (manually adjust the scale IIRC) be incorporated into the code and applied automatically?

**Comment by Desmis** [comment]
@agriggio

No I do not have any, but, reminder, before I saw no differences for the general case.

For the other cases, it is necessary to know that for "edge sharpness", and "denoise refine", 
one uses only the 4 first levels but with a differentiation and important values for each level of contrast.

There are many possible combinations in "edge sharpness", from micro-contrast to sharpening with a completely different approach to "unsharp mask" or "RL deconvolution".

For "final touchup" the algorithm will allow, among other things, to considerably modify the recombination of the contrast to result in something that resembles "Tone mapping"

So there is no reason that there are no differences, but can be like for 3 years now that it satisfies everyone.
It has not disturbed me since the installation of wavelet 2 to 3 years ago.

Jacques

**Comment by Beep6581** [comment]
@agriggio 

> Using the full image when the preview is not 1:1 would slow things down too much

The consensus among people who ask about this is that this would be acceptable if it was a preference option. Personally I would prefer a button which renders a high quality preview when pressed - can't find the issue for that, I'm not sure there is one.

> can the workaround described in rawpedia (manually adjust the scale IIRC) be incorporated into the code and applied automatically?

Unfortunately the workaround was not true, and I removed it from RawPedia a week ago or so.

**Comment by Desmis** [comment]
@agriggio 

The choice between all these propositions is difficult, and reminder, we lived without for a long time.
I think a **button in "wavelet" that would activate "full image" for all wavelet processes** would be an acceptable solution, even if the slowdown would be significant (2 to 5 seconds or more depending on the machine, image, number of levels, ...)

**Comment by Desmis** [comment]
Off topics - I just updated "newwaveletgtk3" with "dev" - no change in code since 9 months.
Recall : sharp-mask and clarity, Retinex in wavelet,  merge files (hdr, style,...)

Jacques

**Comment by luisfl** [comment]
I just observed that the Noise Reduction module in the second tab is also applying partially. So same problem.

EDIT: sorry, I just observed that it shows different results according to different window zoom. So the RT zooming at 200% does not reflect what happens at 100%: it shows more noise reduction that the one actually applied.

Example:

This is what RT shows at 100% (chrominance slider way to the right):
![100](https://cloud.githubusercontent.com/assets/15525268/26431443/23127c72-40cb-11e7-8149-308e14966a3d.jpg)

This is a screenshot of the first picture resized to 4x:
![100resizedto400](https://cloud.githubusercontent.com/assets/15525268/26431441/230de554-40cb-11e7-8e2f-ca5d6366d6a0.jpg)

Yet, resizing to 400% in RT its window shows this:
![400](https://cloud.githubusercontent.com/assets/15525268/26431442/230f6348-40cb-11e7-8b64-c86e33cb3b0f.jpg)

**Comment by agriggio** [comment]
@luisfl for denoising there's an open pr (#3470) that might solve this, maybe you can try it? in the medium term, one possible solution is to have a dedicated "give me an accurate preview" button that trades accuracy for speed

**Comment by luisfl** [comment]
How do I try that? I use the last RPM [here](http://download.opensuse.org/repositories/home:/rawtherapee/Fedora_23/x86_64/). Anyway, from the description does not seem to be related. This is a preview problem.

For this particular scenario, I don't think that a dedicated button would be necessary. If you are zooming to 400% to see the effects of a filter, you are obviously looking for accuracy. Maybe instead of a dedicated button an option in the program Settings?

**Comment by Beep6581** [comment]
@luisfl http://rawpedia.rawtherapee.com/Linux

**Comment by Beep6581** [comment]
Important points:
https://github.com/Beep6581/RawTherapee/issues/3793#issuecomment-290982655
https://github.com/Beep6581/RawTherapee/issues/3793#issuecomment-291078504
https://github.com/Beep6581/RawTherapee/issues/3793#issuecomment-291196488

@heckflosse @agriggio what is the status here - is it fixed, and do we close regardless?

**Comment by agriggio** [comment]
the branch has probably bit-rotten in the meantime. it would take some effort to reintegrate, and I think this is not the way to go. I'd still like to have a unified pipeline, though unfortunately I don't have much time to work on it at the moment

**Comment by Beep6581** [comment]
So, I'll bite the bullet and close this issue. Feel free to re-open if anyone breathes life into this branch.
